### PR TITLE
Feature/unimodal loss

### PR DESCRIPTION
**Pull request by nickgreenspan**
Finished writing and testing batched pytorch implementation of "draw_keypoints" now called "generate_heatmaps" Compared outputs to original numpy function to ensure parity. Implemented the unimodal loss using my new "generate_heatmaps" function, ran tests, and trained model for 11 epochs using it during which the unimodal loss steadily decreased. Note that we are still using the numpy function "draw_keypoints" to generate the ground truth heatmaps for the dataset, could change this later but I wanted to visualize the heatmaps first to double check that my new function is giving the exact same outputs.

### CODE REVIEW

File: tests/test_subpixmaxima.py [hunk 1]
```diff
@@ -49,13 +50,117 @@ def test_subpixmaxima():
...
+        output_sigma=torch.tensor(1.25, device=_TORCH_DEVICE),
+        upsample_factor=torch.tensor(2, device=_TORCH_DEVICE),
+        coordinate_scale=torch.tensor(4, device=_TORCH_DEVICE),  # 2 ** 2
+        confidence_scale=torch.tensor(1, device=_TORCH_DEVICE), 
+        threshold=None,
+        device=_TORCH_DEVICE
+    )
+    og_maxima, gt_confidence = SubPixMax.run(gt_heatmap.to(_TORCH_DEVICE))
+    torch_maxima, confidence_t = SubPixMax.run(torch_heatmap.to(_TORCH_DEVICE))
+    assert(og_maxima == torch_maxima).all()
```
Comment (danbider) [inline tests/test_subpixmaxima.py:93]: nice

File: tests/test_subpixmaxima.py [hunk 2]
```diff
@@ -49,13 +50,117 @@ def test_subpixmaxima():
...
+    assert(gt_confidence == confidence_t).all()
+
+def test_generate_keypoints_weird_shape():
+    data_transform = []
+    OG_SHAPE = (384, 256)
```
Comment (danbider) [inline tests/test_subpixmaxima.py:98]: very good idea

File: tests/test_subpixmaxima.py [hunk 3]
```diff
@@ -49,13 +50,117 @@ def test_subpixmaxima():
...
+    )
+    for batch in dl:
+        _, gt_heatmaps, gt_keypts = batch
+        gt_keypts = gt_keypts.reshape(-1, 17, 2)
+        heatmaps = generate_heatmaps(gt_keypts, height=384, width=384, output_shape=(96,96))
+        gt_maxima, gt_confidence = SubPixMax.run(gt_heatmaps.to(_TORCH_DEVICE))
+        torch_maxima, confidence_t = SubPixMax.run(heatmaps.to(_TORCH_DEVICE))
+        assert(gt_maxima == torch_maxima).all()
+        loss = MaskedMSEHeatmapLoss(gt_heatmaps, heatmaps)
+        assert(loss < 1e-6)
```
Comment (danbider) [inline tests/test_subpixmaxima.py:161]: good.

File: tests/test_subpixmaxima.py [hunk 4]
```diff
@@ -49,13 +50,117 @@ def test_subpixmaxima():
...
+        confidence_scale=torch.tensor(1, device=_TORCH_DEVICE), 
+        threshold=None,
+        device=_TORCH_DEVICE
+    )
+    og_maxima, gt_confidence = SubPixMax.run(gt_heatmap.to(_TORCH_DEVICE))
+    torch_maxima, confidence_t = SubPixMax.run(torch_heatmap.to(_TORCH_DEVICE))
+    assert(og_maxima == torch_maxima).all()
+    assert(gt_confidence == confidence_t).all()
+
+def test_generate_keypoints_batched():
```
Comment (danbider) [inline tests/test_subpixmaxima.py:131]: maybe provide a description about this test? suggestion:
"comparing the ground truth heatmaps generated by numpy with the heatmaps generated from torch. TODO: convert everything to torch".

**Review by danbider — CHANGES_REQUESTED** [review]
@nickgreenspan awesome. 
Added a few suggestions on comments and documentation. Should take a few minutes. then i'll merge.

File: pose_est_nets/losses/losses.py [hunk 5]
```diff
@@ -199,6 +200,23 @@ def TemporalLoss(
...
+def UnimodalLoss(
+    keypoint_preds: TensorType["batch", "two_x_num_keypoints"],
+    heatmap_preds: TensorType["batch", "num_keypoints", "heatmap_height", "heatmap_width"],
+    original_image_height: torch.Tensor,
+    original_image_width: torch.Tensor,
+    output_shape: tuple,
+    **kwargs  # make loss robust to unneeded inputs
+)-> TensorType[(), float]:
+    keypoint_preds = keypoint_preds.reshape(keypoint_preds.shape[0], -1, 2)
+    ideal_heatmaps = generate_heatmaps( #this process doesn't compute gradients
```
[the remark this answers is not in the document]
Comment (nickgreenspan) [reply pose_est_nets/losses/losses.py:213]: I just meant that the process of generating the ideal_heatmaps doesn't compute gradients. As you said this isn't necessary because the backpropagation from the loss goes through heatmap_preds to the weights.

File: pose_est_nets/utils/dataset_utils.py [hunk 6]
```diff
@@ -26,3 +32,39 @@ def draw_keypoints(keypoints, height, width, output_shape, sigma=1, normalize=Tr
...
+    keypoints = keypoints.detach().clone()
+    out_height = output_shape[0]
+    out_width = output_shape[1]
+    keypoints[:, :, 1] *= out_height / height
+    keypoints[:, :, 0] *= out_width / width
+    nan_idxes = torch.isnan(keypoints)[:, :, 0]
+    #confidence = torch.zeros((batch_dim, n_keypoints, out_height, out_width), device=keypoints.device)
+    xv = torch.arange(out_width, device=keypoints.device)
+    yv = torch.arange(out_height, device=keypoints.device)
+    xx, yy = torch.meshgrid(yv, xv) #double check flipped order because of pytorch's ij and numpy's xy indexing for meshgrid
```
[the remark this answers is not in the document]
Comment (nickgreenspan) [reply pose_est_nets/utils/dataset_utils.py:56]: Yes, I checked for that.

File: pose_est_nets/utils/heatmap_tracker_utils.py [hunk 7]
```diff
@@ -238,36 +238,24 @@ def __init__(
 
     @property
     def kernel_size(self):
-        kernel_size = np.min(self.output_shape)
+        kernel_size = np.min(self.output_shape) #could change to torch.min
```
[the remark this answers is not in the document]
Comment (nickgreenspan) [reply pose_est_nets/utils/heatmap_tracker_utils.py:241]: I'm going to keep it as is for now, torch.min doesn't work on python tuples but np.min does.

**Review by danbider — APPROVED** [review]
